File: ai-demo-3/backend/app/local_db.py
"""
Local SQLite Database Module
Provides same interface as redshift.py but uses SQLite.
"""
import sqlite3
import logging
from typing import Dict, List, Optional, Any
import os

logger = logging.getLogger(__name__)

DB_PATH = os.path.join(os.path.dirname(__file__), "..", "local_crm.db")

def get_connection():
    """Get SQLite connection."""
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

async def lookup_hcp_by_name(name: str) -> Optional[Dict[str, Any]]:
    """Look up HCP by name (case-insensitive)."""
    try:
        conn = get_connection()
        cursor = conn.cursor()
        
        cursor.execute(
            "SELECT hcp_id, name, hco_id FROM hcp WHERE LOWER(name) = LOWER(?)",
            (name,)
        )
        
        row = cursor.fetchone()
        conn.close()
        
        if row:
            return {
                "hcp_id": row["hcp_id"],
                "name": row["name"],
                "hco_id": row["hco_id"],
                "found": True
            }
        
        return None
        
    except Exception as e:
        logger.error(f"Error looking up HCP: {e}")
        return None

async def insert_call(record: Dict[str, Any]) -> str:
    """Insert call record into SQLite."""
    import uuid
    
    call_pk = f"CALL_{uuid.uuid4().hex[:12].upper()}"
    
    try:
        conn = get_connection()
        cursor = conn.cursor()
        
        # Extract follow-up task
        followup = record.get('call_follow_up_task', {})
        
        cursor.execute("""
            INSERT INTO calls (
                call_pk, call_channel, discussion_topic, status, account, id,
                adverse_event, adverse_event_details,
                noncompliance_event, noncompliance_description,
                call_notes, call_date, call_time, product,
                followup_task_type, followup_description, followup_due_date, followup_assigned_to,
                created_at
            ) VALUES (
                ?, ?, ?, ?, ?, ?,
                ?, ?,
                ?, ?,
                ?, ?, ?, ?,
                ?, ?, ?, ?,
                CURRENT_TIMESTAMP
            )
        """, (
            call_pk,
            record.get('call_channel', 'In-person'),
            record.get('discussion_topic', ''),
            record.get('status', 'Saved_vod'),
            record.get('account', ''),
            record.get('id', ''),
            record.get('adverse_event', False),
            record.get('adverse_event_details', ''),
            record.get('noncompliance_event', False),
            record.get('noncompliance_description', ''),
            record.get('call_notes', ''),
            record.get('call_date', ''),
            record.get('call_time', ''),
            record.get('product', ''),
            followup.get('task_type', ''),
            followup.get('description', ''),
            followup.get('due_date', ''),
            followup.get('assigned_to', '')
        ))
        
        conn.commit()
        conn.close()
        
        logger.info(f"✅ Call record inserted: {call_pk}")
        return call_pk
        
    except Exception as e:
        logger.error(f"Failed to insert call: {e}")
        raise

async def get_recent_calls(limit: int = 10) -> List[Dict[str, Any]]:
    """Get recent calls."""
    try:
        conn = get_connection()
        cursor = conn.cursor()
        
        cursor.execute(
            "SELECT * FROM calls ORDER BY created_at DESC LIMIT ?",
            (limit,)
        )
        
        rows = cursor.fetchall()
        conn.close()
        
        return [dict(row) for row in rows]
        
    except Exception as e:
        logger.error(f"Error fetching calls: {e}")
        return []

async def check_connection() -> bool:
    """Check database connection."""
    try:
        conn = get_connection()
        conn.close()
        return True
    except Exception as e:
        logger.error(f"Database check failed: {e}")
        return False

File: ai-demo-3-nodejs/src/localDb.ts
/**
 * Local SQLite Database Module
 * Provides same interface as redshift.ts but uses SQLite
 */
import sqlite3 from 'sqlite3';
import { open, Database } from 'sqlite';
import * as path from 'path';
import logger from './logger';

const DB_PATH = path.join(__dirname, '..', 'local_crm.db');

let db: Database | null = null;

/**
 * Get database connection
 */
async function getConnection(): Promise<Database> {
  if (!db) {
    db = await open({
      filename: DB_PATH,
      driver: sqlite3.Database,
    });
  }
  return db;
}

/**
 * Look up HCP by name (case-insensitive)
 */
export async function lookupHcpByName(name: string): Promise<any | null> {
  try {
    const conn = await getConnection();
    
    const row = await conn.get(
      'SELECT hcp_id, name, hco_id FROM hcp WHERE LOWER(name) = LOWER(?)',
      [name]
    );
    
    if (row) {
      return {
        hcp_id: row.hcp_id,
        name: row.name,
        hco_id: row.hco_id,
        found: true,
      };
    }
    
    return null;
  } catch (error) {
    logger.error('Error looking up HCP:', error);
    return null;
  }
}

/**
 * Insert call record into SQLite
 */
export async function insertCall(record: any): Promise<string> {
  const callPk = `CALL_${Date.now()}_${Math.random().toString(36).substring(7)}`;
  
  try {
    const conn = await getConnection();
    
    // Extract follow-up task
    const followup = record.call_follow_up_task || {};
    
    await conn.run(
      `INSERT INTO calls (
        call_pk, call_channel, discussion_topic, status, account, id,
        adverse_event, adverse_event_details,
        noncompliance_event, noncompliance_description,
        call_notes, call_date, call_time, product,
        followup_task_type, followup_description, followup_due_date, followup_assigned_to,
        created_at
      ) VALUES (
        ?, ?, ?, ?, ?, ?,
        ?, ?,
        ?, ?,
        ?, ?, ?, ?,
        ?, ?, ?, ?,
        CURRENT_TIMESTAMP
      )`,
      [
        callPk,
        record.call_channel || 'In-person',
        record.discussion_topic || '',
        record.status || 'Saved_vod',
        record.account || '',
        record.id || '',
        record.adverse_event || false,
        record.adverse_event_details || '',
        record.noncompliance_event || false,
        record.noncompliance_description || '',
        record.call_notes || '',
        record.call_date || '',
        record.call_time || '',
        record.product || '',
        followup.task_type || '',
        followup.description || '',
        followup.due_date || '',
        followup.assigned_to || '',
      ]
    );
    
    logger.info(`✅ Call record inserted: ${callPk}`);
    return callPk;
  } catch (error) {
    logger.error('Failed to insert call:', error);
    throw error;
  }
}

/**
 * Get recent calls
 */
export async function getRecentCalls(limit: number = 10): Promise<any[]> {
  try {
    const conn = await getConnection();
    
    const rows = await conn.all(
      'SELECT * FROM calls ORDER BY created_at DESC LIMIT ?',
      [limit]
    );
    
    return rows;
  } catch (error) {
    logger.error('Error fetching calls:', error);
    return [];
  }
}

/**
 * Check database connection
 */
export async function checkConnection(): Promise<boolean> {
  try {
    await getConnection();
    return true;
  } catch (error) {
    logger.error('Database check failed:', error);
    return false;
  }
}

/**
 * Generic query method (PostgreSQL-compatible interface)
 * Converts SQLite results to match Redshift format
 */
export async function query(sql: string, params: any[] = []): Promise<{ rows: any[] }> {
  try {
    const conn = await getConnection();
    
    // Convert PostgreSQL $1, $2 parameters to SQLite ? placeholders
    let sqliteSql = sql;
    let sqliteParams = params;
    
    // Check if using PostgreSQL-style parameters ($1, $2, etc.)
    if (sql.includes('$')) {
      // Replace $1, $2, etc. with ?
      sqliteSql = sql.replace(/\$\d+/g, '?');
    }
    
    // Determine if this is a SELECT query
    const isSelect = sqliteSql.trim().toUpperCase().startsWith('SELECT');
    
    if (isSelect) {
      const rows = await conn.all(sqliteSql, sqliteParams);
      return { rows };
    } else {
      // INSERT/UPDATE/DELETE
      const result = await conn.run(sqliteSql, sqliteParams);
      return {
        rows: [{ 
          rowCount: result.changes,
          lastID: result.lastID 
        }]
      };
    }
  } catch (error) {
    logger.error('SQLite query failed:', error);
    throw error;
  }
}

/**
 * Check if database is available
 */
export function isAvailable(): boolean {
  return true; // SQLite is always available locally
}

// Export default object with all methods
export default {
  lookupHcpByName,
  insertCall,
  getRecentCalls,
  checkConnection,
  query,
  isAvailable,
};


